File: v4/rust/iprange-livedb/src/validation/structure_table.rs
//! Validation traversal for the direct structure-ID table.

use crate::error::{Error, Result};
use crate::mapping::{ByteRange, ByteSource, PageView};
use crate::structured_value::{table, PayloadCodec};

use super::context::Context;
use super::{ValidationObject, ValidationReason, ValidationSink};

const MAX_DEPTH: usize = 4;

pub(crate) struct WalkResult {
    pub(crate) records: u64,
}

pub(crate) fn walk<'m, P, S, F>(
    context: &mut Context<'m, S>,
    root: u32,
    mut leaf: F,
) -> Result<WalkResult>
where
    P: PayloadCodec,
    S: ValidationSink,
    F: FnMut(&mut Context<'m, S>, u32, u64, ByteRange<PageView<'m>>) -> Result<()>,
{
    if root == 0 {
        return Ok(WalkResult { records: 0 });
    }
    let level = table::required_level::<P>(context.meta.structure_id_limit)?;
    let mut path = [0; MAX_DEPTH];
    let mut records = 0u64;
    walk_node::<P, S, F>(
        context,
        root,
        level,
        0,
        &mut path,
        0,
        &mut records,
        &mut leaf,
    )?;
    Ok(WalkResult { records })
}

#[allow(clippy::too_many_arguments)]
fn walk_node<'m, P, S, F>(
    context: &mut Context<'m, S>,
    page_number: u32,
    expected_level: u16,
    base: u64,
    path: &mut [u32; MAX_DEPTH],
    depth: usize,
    records: &mut u64,
    leaf: &mut F,
) -> Result<()>
where
    P: PayloadCodec,
    S: ValidationSink,
    F: FnMut(&mut Context<'m, S>, u32, u64, ByteRange<PageView<'m>>) -> Result<()>,
{
    let Some(slot) = path.get_mut(depth) else {
        return context.emit(
            ValidationReason::TreeLevelInvalid,
            ValidationObject::StructureDictionary,
            Some(page_number),
            None,
            None,
        );
    };
    *slot = page_number;
    let Some(page) = context.read_graph_page(
        page_number,
        ValidationObject::StructureDictionary,
        &path[..depth],
    )?
    else {
        return Ok(());
    };
    let header =
        match table::inspect_header::<P, _>(page, context.meta.txn_id, Some(expected_level)) {
            Ok(header) => header,
            Err(problem) => {
                return context.emit(
                    header_reason(problem),
                    ValidationObject::StructureDictionary,
                    Some(page_number),
                    None,
                    None,
                );
            }
        };
    if !table::reserved_zero::<P, _>(page, header.level) {
        context.emit(
            ValidationReason::PageReservedNonzero,
            ValidationObject::StructureDictionary,
            Some(page_number),
            None,
            None,
        )?;
    }
    if header.level == 0 {
        walk_leaf::<P, S, F>(context, page_number, page, base, header, records, leaf)
    } else {
        walk_branch::<P, S, F>(
            context,
            page_number,
            page,
            base,
            header,
            path,
            depth,
            records,
            leaf,
        )
    }
}

fn walk_leaf<'m, P, S, F>(
    context: &mut Context<'m, S>,
    page_number: u32,
    page: PageView<'m>,
    base: u64,
    header: table::Header,
    records: &mut u64,
    leaf: &mut F,
) -> Result<()>
where
    P: PayloadCodec,
    S: ValidationSink,
    F: FnMut(&mut Context<'m, S>, u32, u64, ByteRange<PageView<'m>>) -> Result<()>,
{
    let mut found = 0usize;
    for slot in 0..table::leaf_slots::<P>() {
        context.checkpoint()?;
        let cell = table::leaf_cell::<P, _>(page, slot)?;
        if cell.all_zero(0, cell.len()) {
            continue;
        }
        found += 1;
        *records = records.checked_add(1).ok_or(Error::ArithmeticOverflow(
            "structure validation record count",
        ))?;
        leaf(context, page_number, base + slot as u64, cell)?;
    }
    if found != header.item_count {
        page_shape_finding(context, page_number)?;
    }
    Ok(())
}

#[allow(clippy::too_many_arguments)]
fn walk_branch<'m, P, S, F>(
    context: &mut Context<'m, S>,
    page_number: u32,
    page: PageView<'m>,
    base: u64,
    header: table::Header,
    path: &mut [u32; MAX_DEPTH],
    depth: usize,
    records: &mut u64,
    leaf: &mut F,
) -> Result<()>
where
    P: PayloadCodec,
    S: ValidationSink,
    F: FnMut(&mut Context<'m, S>, u32, u64, ByteRange<PageView<'m>>) -> Result<()>,
{
    let span = table::coverage::<P>(header.level - 1)?;
    let mut found = 0usize;
    for index in 0..table::branch_slots() {
        context.checkpoint()?;
        let child = table::raw_branch_child(page, index)?;
        if child == 0 {
            continue;
        }
        found += 1;
        let child_base = base
            .checked_add(
                span.checked_mul(index as u64)
                    .ok_or(Error::ArithmeticOverflow("structure validation coverage"))?,
            )
            .ok_or(Error::ArithmeticOverflow("structure validation coverage"))?;
        walk_node::<P, S, F>(
            context,
            child,
            header.level - 1,
            child_base,
            path,
            depth + 1,
            records,
            leaf,
        )?;
    }
    if found != header.item_count {
        page_shape_finding(context, page_number)?;
    }
    Ok(())
}

fn header_reason(problem: table::HeaderProblem) -> ValidationReason {
    match problem {
        table::HeaderProblem::Header | table::HeaderProblem::Shape => {
            ValidationReason::PageHeaderInvalid
        }
        table::HeaderProblem::Born => ValidationReason::PageBornTxnInvalid,
        table::HeaderProblem::Type => ValidationReason::PageTypeMismatch,
        table::HeaderProblem::Level => ValidationReason::TreeLevelInvalid,
    }
}

fn page_shape_finding<S: ValidationSink>(
    context: &mut Context<'_, S>,
    page_number: u32,
) -> Result<()> {
    context.emit(
        ValidationReason::PageHeaderInvalid,
        ValidationObject::StructureDictionary,
        Some(page_number),
        None,
        None,
    )
}

File: v4/go/internal/validation/structure_table.go
package validation

// Structure-ID table validation walk (Rust validation/structure_table.rs
// over the structured_value/table.rs geometry): the dense radix table
// walk with the directory/record page header inspection, the per-slot
// record discovery, the found/item-count proof, and the reserved-tail
// scan. The walk is allocation-free and follows the graph claims; the
// record cells alias the inspected page.

import (
	"github.com/firehol/iprange/v4/go/internal/format"
)

// structureTableMaxDepth mirrors Rust MAX_DEPTH: one path slot per
// level, for at most the level-3 directory root plus its record page.
const structureTableMaxDepth = 4

// structureTableWalkResult mirrors Rust WalkResult.
type structureTableWalkResult struct {
	records uint64
}

// walkStructureTable mirrors Rust structure_table::walk over the root:
// a zero root is an empty table, every nonzero record slot runs leaf
// with its implied id, and the result carries the decoded record count.
func walkStructureTable(ctx *context, root uint32, leaf func(*context, uint32, uint64, []byte) error) (structureTableWalkResult, error) {
	if root == 0 {
		return structureTableWalkResult{}, nil
	}
	level, err := structureTableRequiredLevel(ctx.meta.StructureIDLimit)
	if err != nil {
		return structureTableWalkResult{}, err
	}
	var path [structureTableMaxDepth]uint32
	records := uint64(0)
	if err := walkStructureTableNode(ctx, root, level, 0, &path, 0, &records, leaf); err != nil {
		return structureTableWalkResult{}, err
	}
	return structureTableWalkResult{records: records}, nil
}

// structureTableRequiredLevel mirrors structured_value/table.rs
// required_level: the ID limit must fit the u32 namespace and the
// smallest level whose coverage reaches it is the root level.
func structureTableRequiredLevel(limit uint64) (uint16, error) {
	if limit < 1 || limit > uint64(1)<<32 {
		return 0, &format.Error{Code: format.CodeFormatInvalid, Detail: "structure table ID limit is invalid"}
	}
	level, ok := format.StructureRootLevel(limit)
	if !ok {
		return 0, &format.Error{Code: format.CodeFormatInvalid, Detail: "structure table ID limit is invalid"}
	}
	return uint16(level), nil
}

// walkStructureTableNode visits one dense table node (Rust walk_node):
// the path slot, the graph page, the classified header, the reserved
// tail, and the leaf or directory walk.
func walkStructureTableNode(ctx *context, pageNumber uint32, expectedLevel uint16, base uint64, path *[structureTableMaxDepth]uint32, depth int, records *uint64, leaf func(*context, uint32, uint64, []byte) error) error {
	if depth >= len(path) {
		pageCopy := pageNumber
		if err := ctx.emit(ReasonTreeLevelInvalid, ObjectStructureDictionary, &pageCopy, nil, nil); err != nil {
			return err
		}
		return nil
	}
	path[depth] = pageNumber
	page, err := ctx.readGraphPage(pageNumber, ObjectStructureDictionary, path[:depth])
	if err != nil || page == nil {
		return err
	}
	header, ok, err := structureTablePageHeader(ctx, pageNumber, page, &expectedLevel)
	if err != nil || !ok {
		return err
	}
	if !format.AllZero(page[header.Lower:format.PageSize]) {
		pageCopy := pageNumber
		if err := ctx.emit(ReasonPageReservedNonzero, ObjectStructureDictionary, &pageCopy, nil, nil); err != nil {
			return err
		}
	}
	if header.Level == 0 {
		return walkStructureTableLeaf(ctx, pageNumber, page, base, header, records, leaf)
	}
	return walkStructureTableBranch(ctx, pageNumber, page, base, header, path, depth, records, leaf)
}

// structureTablePageHeader runs the dense-table header inspection and
// streams the classified finding on refusal (Rust structure_table
// header_reason: the same classes as the tree pages).
func structureTablePageHeader(ctx *context, pageNumber uint32, page []byte, expectedLevel *uint16) (format.PageHeader, bool, error) {
	header, problem := format.InspectStructureTableHeader(page, ctx.meta.TxnID, uint32(ctx.meta.StructureKind), expectedLevel)
	if problem != format.TreeHeaderProblemNone {
		pageCopy := pageNumber
		if err := ctx.emit(treeHeaderProblemReason(problem), ObjectStructureDictionary, &pageCopy, nil, nil); err != nil {
			return format.PageHeader{}, false, err
		}
		return format.PageHeader{}, false, nil
	}
	return header, true, nil
}

// walkStructureTableLeaf visits every nonzero record slot (Rust
// walk_leaf): the slot cells in id order, the record count, and the
// leaf callback with the implied id base+slot.
func walkStructureTableLeaf(ctx *context, pageNumber uint32, page []byte, base uint64, header format.PageHeader, records *uint64, leaf func(*context, uint32, uint64, []byte) error) error {
	found := 0
	for slot := 0; slot < format.StructureRecordSlots; slot++ {
		if err := ctx.checkpoint(); err != nil {
			return err
		}
		cell := page[32+slot*format.StructureRecordSize : 32+(slot+1)*format.StructureRecordSize]
		if format.AllZero(cell) {
			continue
		}
		found++
		if *records == ^uint64(0) {
			return &format.Error{Code: format.CodeArithmeticOverflow, Detail: "validation structure record count"}
		}
		*records++
		if err := leaf(ctx, pageNumber, base+uint64(slot), cell); err != nil {
			return err
		}
	}
	if found != int(header.ItemCount) {
		return structureTableShapeFinding(ctx, pageNumber)
	}
	return nil
}

// walkStructureTableBranch visits every nonzero directory child (Rust
// walk_branch): the child span at the level below, the child base
// arithmetic, and the recursion with the next lower level.
func walkStructureTableBranch(ctx *context, pageNumber uint32, page []byte, base uint64, header format.PageHeader, path *[structureTableMaxDepth]uint32, depth int, records *uint64, leaf func(*context, uint32, uint64, []byte) error) error {
	span, ok := format.StructureSpanOfLevel(uint32(header.Level))
	if !ok {
		return &format.Error{Code: format.CodeArithmeticOverflow, Detail: "validation structure coverage"}
	}
	found := 0
	for index := 0; index < format.StructureDirectoryChildCount; index++ {
		if err := ctx.checkpoint(); err != nil {
			return err
		}
		child := format.U32(page[32+index*4 : 36+index*4])
		if child == 0 {
			continue
		}
		found++
		offset := span * uint64(index)
		if index != 0 && offset/uint64(index) != span {
			return &format.Error{Code: format.CodeArithmeticOverflow, Detail: "validation structure coverage"}
		}
		childBase := base + offset
		if childBase < base {
			return &format.Error{Code: format.CodeArithmeticOverflow, Detail: "validation structure coverage"}
		}
		if err := walkStructureTableNode(ctx, child, header.Level-1, childBase, path, depth+1, records, leaf); err != nil {
			return err
		}
	}
	if found != int(header.ItemCount) {
		return structureTableShapeFinding(ctx, pageNumber)
	}
	return nil
}

// structureTableShapeFinding streams the header class of a record or
// directory page whose found cells disagree with its item count (Rust
// page_shape_finding).
func structureTableShapeFinding(ctx *context, pageNumber uint32) error {
	pageCopy := pageNumber
	return ctx.emit(ReasonPageHeaderInvalid, ObjectStructureDictionary, &pageCopy, nil, nil)
}
